### ISSUE TITLE
[FR] Autologging functionality for PyTorch/fastai

### ISSUE THREAD

**Issue by brookewenig** [mlflow/mlflow#2087]
## Describe the proposal

It'd be nice to add an mlflow.pytorch.autolog() API for automatically logging metrics, params & models generated via PyTorch. In particular, I'd like this to be incorporated with the fastai library (built on PyTorch), which also generates artifacts such as learning rate finder plots, and autolog those as well. 

### Motivation
PyTorch is heavily used in research for Deep Learning and is becoming very prevalent in industry as well with the fastai library. The fastai community has over 450 contributors, and it incorporates practical deep learning best practices + cutting edge algorithms in the library. In the fast.ai course that is run by USF, they use Excel spreadsheets to keep track of their runs. If we could just autolog everything once they import mlflow, it will help bring the adoption of MLflow to the fastai community. 

### Proposed Changes
See above - we propose adding a new mlflow.pytorch.autolog API

**Comment by smurching** [comment]
@shalomma @LarsHill would either of you be interested in working on this feature? If so I'm happy to write up / help collaborate on an API proposal :)

**Comment by LarsHill** [comment]
@smurching sorry, but in the near future I won't have time working on it. I just thought it would be a useful feature.

**Comment by pierrelandry** [comment]
@smurching are you working on it ? I will be happy to help :)

**Comment by smurching** [comment]
@pierrelandry I haven't worked on it, that'd be awesome! Before writing code & opening a PR, a good first step would be to write up a description of the proposed API & its behavior. Happy to review a proposal if you're able to write one & share it here.

I'm new to Pytorch, but I'd recommend focusing on adding autologging for the most common Pytorch APIs first (will spend some time tomorrow learning more about Pytorch). Also, I'd suggest `mlflow.pytorch.autolog()` as the API name, to match the existing Keras & TF autologging APIs, but open to alternatives. Thanks for being willing to contribute :)!

**Comment by smurching** [comment]
@pierrelandry hope you enjoyed the holidays! Thanks again for your interest in helping with this - are you still interested in collaborating?

**Comment by antoniomdk** [comment]
Hi, I'd love to work on the fastai integration, is there any proposal API?
I thought of adding support for logging metrics during training and saving artifacts generated from
[SaveModelCallback](https://docs.fast.ai/callbacks.html#SaveModelCallback) or [Learner.save](https://docs.fast.ai/basic_train.html#Learner.save) method.

**Comment by smurching** [comment]
@antoniomdk sorry for the delayed response, that'd be awesome - one strawman API proposal would be to add an `mlflow.pytorch.autolog()` API, and an `mlflow.fastai.autolog()` if appropriate - AFAICT fastai & PyTorch are analogous to Keras and TensorFlow (fastai seems to be a higher-level API implemented using PyTorch), so seems reasonable to have separate APIs for the two libraries. 

That being said, it's probably worth considering whether implementing PyTorch autologging could indirectly solve fastai autologging (i.e. could we reuse the implementation of PyTorch autologging for fastai autologging?). My guess is that it might be possible, but the right way to implement fastai autologging would be through public `fastai` APIs, e.g. by adding callbacks to save & log the user's model as you described.

**Comment by antoniomdk** [comment]
Yes, you're right, Fastai is a higher-level API on top of Pytorch as Keras is on top of Tensorflow. Adding model saving autologging for Pytorch would solve it also for Fastai, as the latter relies on torch.save
https://github.com/fastai/fastai/blob/master/fastai/basic_train.py#L246
So, monkey patching [torch.save](https://pytorch.org/docs/stable/torch.html?highlight=save#torch.save) would work for both libraries. It'd also work for model checking along epochs. 

On the other hand, I think we should go for adding a callback to handle metric logging on Fastai.
There is also a proposal from the Fastai community to add a callback for this: https://forums.fast.ai/t/mlflow-tracking-support/38893

For Pytorch I'm not sure how we should proceed because the training cycle is done manually and the evaluation depends on the user implementing it.

**Comment by jacobhess118** [comment]
Adding autologging to PyTorch is much harder inherently because PyTorch's training cycle can be very flexible and can take many forms, like @antoniomdk mentioned. I took a look at what MLFlow autologging does for other frameworks (https://www.mlflow.org/docs/latest/tracking.html#automatic-logging-from-tensorflow-and-keras-experimental), and I think we might be able to do the following:

- Metrics

| Logging Item | Possible Implementation |
| ------------- | ------------- |
| training loss  | Monkey-patch [Tensor.backward](https://pytorch.org/docs/stable/autograd.html#torch.Tensor.backward) method and log the tensor |
| validation loss  | (This is difficult to track because we can't know programmatically which variable is the validation loss. We might have to resort to manual logging.) |
| user-specified metrics | (Not sure how we should map this concept from Keras to PyTorch.) |

- Parameters

| Logging Item | Possible Implementation |
| ------------- | ------------- |
| fit() parameters  | (Not sure how we should map this concept from Keras to PyTorch.) |
| optimizer name  | Find instantiations of [torch.optim.Optimizer](https://pytorch.org/docs/stable/optim.html#torch.optim.Optimizer) subclasses in the code, and log the optimizer name |
| learning rate | Log the `lr` value of the optimizer found above |
| epsilon | Log the `eps` value of the optimizer found above |

- Artifacts

| Logging Item | Possible Implementation |
| ------------- | ------------- |
| model summary on training start | Monkey-patch `Tensor.backward` or `Optimizer.step` method and log the model summary on the first method call |
| MLFlow model (PyTorch model) on training end | The end of training is hard to detect, but maybe we can monkey-patch `Optimizer.step` or `torch.nn.Module.eval` method to trigger MLFlow model saving after every epoch |

**Comment by jacobhess118** [comment]
> AFAICT fastai & PyTorch are analogous to Keras and TensorFlow (fastai seems to be a higher-level API implemented using PyTorch), so seems reasonable to have separate APIs for the two libraries.

+1 I think it's easier to add autologging to fastai as a separate API because callbacks are a cleaner way to implement autologging.

**Comment by antoniomdk** [comment]
I think it'd be hard to detect the last epoch using Optimizer.step or torch.nn.Module.eval.
For model persistence it's easier to rely on [torch.save](https://pytorch.org/docs/stable/torch.html?highlight=save#torch.save).

For optimizer tracking we can monkey-patch torch.optim.Optimizer class to keep track of all instances created. We can use those instances to log optimizer parameters, as @jacobhess118 suggested.

For tracking loss I'm not sure if using Tensor.backward would be a good idea (could it make Tensor.backward subcalls?).

**Comment by smurching** [comment]
@jacobhess118 @antoniomdk thanks for the detailed investigation! Here are my impressions:
* Makes sense to implement fastai autologging via a separate `mlflow.fastai.autolog` API, where calling `mlflow.fastai.autolog` patches [`Learner.fit`](https://docs.fast.ai/basic_train.html#Learner.fit) & modifies the `callbacks` argument, injecting a callback that logs model & training information (see [analogous implementation](https://github.com/mlflow/mlflow/blob/1a762d56564f5f708bd7f9cbc5a48530fb354ae3/mlflow/keras.py#L528) in Keras autologging)
* PyTorch autologging: PyTorch is a popular framework (2x the GitHub stars of fastai) so would be nice to have autologging support, but if it seems likely to be brittle/complex (as it currently seems from @jacobhess118's description), we can punt on it for now - it's possible it'll become easier down the line

Thus @antoniomdk if you're interested in adding `fastai` autologging for now, would be happy to review a PR that does so

**Comment by jacobhess118** [comment]
> For tracking loss I'm not sure if using Tensor.backward would be a good idea (could it make Tensor.backward subcalls?).

Tensor.backward doesn’t make subcalls of itself at the moment (and likely won’t be in the future) because calling Python functions is slow. However I do think that it smells more like a hack, and we probably shouldn’t do it this way unless absolutely necessary.

**Comment by smurching** [comment]
Hi @antoniomdk, just wanted to check in (no rush of course!) - did you get a chance to get started with this and/or have any questions?

**Comment by antoniomdk** [comment]
Hi @smurching, I started to work on this recently. I've forked the repo and create a new branch but my advances are slow. I'm quite busy at this moment at work, but I hope to make a PR by the end of this month.

**Comment by smurching** [comment]
Awesome - totally makes sense & sounds good! Let us know if anything comes up along the way :)

**Comment by antoniomdk** [comment]
Sorry for the late response, but I just want you to know that I'm working on the PR. I've already got cover most of the functionality for autologging and model exporting of Fastai models.

I'm figuring out until which extend should we support autologging for specific callbacks.
There is a callback for saving the best model of the training, Early Stopping, etc.
So, I'm not sure how to proceed with those callbacks. Should we support just Early Stopping?
Or should we add [SaveModelCallback](SaveModelCallback) too?

I've added a custom callback to save the model params before training and the model itself after training. We can use that callback to track the best model along epochs.

What do you think guys?

**Comment by antoniomdk** [comment]
I've created a PR for this @smurching . It has only support for logging Early Stopping and OneCycleScheduler params (along with the other .fit params).
